rig-neo4j/src/lib.rs

```diff
@@ -0,0 +1,297 @@
+use neo4rs::*;
+use std::fmt::Debug;
+
+use rig::{
+    embeddings::{DocumentEmbeddings, Embedding, EmbeddingModel},
+    vector_store::{VectorStore, VectorStoreError, VectorStoreIndex},
+};
+use serde::Deserialize;
+
+pub struct Neo4jClient {
+    pub graph: Graph,
+}
+
+/// A Neo4j vector store.
+pub struct Neo4jVectorStore {
+    //collection: mongodb::Collection<DocumentEmbeddings>,
+    pub client: Neo4jClient,
+    pub database_name: String,
+}
+
+fn neo4j_to_rig_error(e: neo4rs::Error) -> VectorStoreError {
+    VectorStoreError::DatastoreError(Box::new(e))
+}
+
+trait ToBoltType {
+    fn to_bolt_type(&self) -> BoltType;
+}
+
+impl ToBoltType for Embedding {
+    fn to_bolt_type(&self) -> BoltType {
+        let mut bolt_map = BoltMap::new();
+        let bolt_list = BoltType::List(
+            self.vec
+                .iter()
+                .map(|&f| BoltType::Float(BoltFloat::new(f)))
+                .collect::<Vec<BoltType>>()
+                .into(),
+        );
+        bolt_map.put(BoltString::new(&self.document), bolt_list);
+        BoltType::Map(bolt_map)
+    }
+}
+
+impl Neo4jClient {
+    pub fn new(graph: Graph) -> Self {
+        Self { graph }
+    }
+
+    pub async fn connect(uri: &str, user: &str, password: &str) -> Result<Self, VectorStoreError> {
+        let graph = Graph::new(uri, user, password)
+            .await
+            .map_err(|e| VectorStoreError::DatastoreError(Box::new(e)))?;
+        Ok(Self { graph })
+    }
+}
+
+impl VectorStore for Neo4jVectorStore {
+    type Q = neo4rs::Query;
+
+    async fn add_documents(
+        &mut self,
+        documents: Vec<DocumentEmbeddings>,
+    ) -> Result<(), VectorStoreError> {
+        for doc in documents {
+            let query = Query::new("CREATE (d:DocumentEmbeddings {id: $id, document: $document, embeddings: $embeddings})".to_string())
```

Comment:
Maybe use a const for premade queries.

Reply:
This will be removed as the VectorStore trait is no more 🪦.